### PR TITLE
Pull Request PayTrace

### PR DESCRIPTION
**Pull request by xlreconlx**

### CODE REVIEW

**Comment by jonesde** [comment]
As I mentioned in the pull request for moqui-framework to add another variation on the simple HTTP request methods: better to use the RestClient (see interfaces in RestClient.java). The best examples of its use are in ShippoServices.xml. This is far better for flexibility, error handling, parsing/generating JSON, etc.

File: service/PayTrace/PayTraceServices.xml [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            <service-call name="create#mantle.account.method.PaymentGatewayResponse" out-map="context"
+                    in-map="[paymentGatewayConfigId:paymentGatewayConfigId, paymentOperationEnumId:'PgoStoreCim',
+                            paymentMethodId:paymentMethodId,paymentId:paymentId, transactionDate:ec.user.nowTimestamp,
+                            resultSuccess:'Y', resultError:'N', resultBadCardNumber:'N', referenceNum: '']"/>
+
+            <!-- set the gatewayCimId values for Party and PaymentMethod as needed -->
+            <if condition="paymentMethod.gatewayCimId"><then>
+                <!-- have a payment profile, called updateCustomerPaymentProfileRequest, no need to do anything -->
+            </then><else-if condition="party.gatewayCimId">
+                <!-- no payment profile but we have a customer profile, called createCustomerPaymentProfileRequest -->
```
Comment (danieltaylor-nz) [inline service/PayTrace/PayTraceServices.xml]: these comments are from Authorize.net services, should be removed or updated for Paytrace.

File: service/PayTrace/PayTracePaymentServices.xml [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+                <if condition="payment == null"><return message="Payment ${paymentId} not found"/></if>
+                <set field="paymentMethod" from="payment.'mantle.account.method.PaymentMethod'"/>
+                <if condition="paymentMethod?.paymentMethodTypeEnumId != 'PmtCreditCard'">
+                    <return error="true" message="Cannot capture payment ${paymentId}, not a Credit Card payment."/>
+                </if>
+
+                <!-- if no gatewayCimId, store the PaymentMethod on CIM -->
+                <if condition="!paymentMethod.gatewayCimId">
+                    <service-call name="AuthorizeDotNet.CimServices.store#CustomerPaymentMethod"
+                        in-map="[paymentMethodId:paymentMethod.paymentMethodId, paymentId:paymentId]"/>
```
Comment (danieltaylor-nz) [inline service/PayTrace/PayTracePaymentServices.xml]: I think you mean:
`<service-call name="PayTrace.PayTraceServices.store#CustomerPaymentMethod"
                            in-map="[paymentMethodId:paymentMethod.paymentMethodId, paymentId:paymentId, paymentGatewayConfigId:paymentGatewayConfigId]"/>`

File: service/PayTrace/PayTracePaymentServices.xml [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+                    <set field="cardSecurityCode" from="creditCard?.cardSecurityCode"/>
+                </if>
+                <if condition="cardSecurityCode == null"><return error="true" message="Cannot authorize payment ${paymentId}, must include a cardSecurityCode."/></if>
+
+                <set field="postalAddress" from="paymentMethod.'mantle.party.contact.PostalAddress'"/>
+               
+                <!-- assemble the full request Map -->
+                <set field="requestMap" from="[amount:payment.amount,credit_card:[number:creditCard?.cardNumber,expiration_month:creditCard?.expireDate ? creditCard.expireDate.substring(0, creditCard.expireDate.indexOf('/')) : null,expiration_year:creditCard?.expireDate ? creditCard.expireDate.substring(creditCard.expireDate.indexOf('/')+1],csc:cardSecurityCode ? cardSecurityCode: '',billing_address:[name:postalAddress?.toName,street_address:postalAddress?.address1,city:postalAddress?.city,state:postalAddress?.stateProvinceGeoId ? postalAddress.stateProvinceGeoId.substring(postalAddress.stateProvinceGeoId.indexOf('_')+1) : null,zip:postalAddress?.postalCode]]"/>
```
Comment (danieltaylor-nz) [inline service/PayTrace/PayTracePaymentServices.xml]: spaces after commas:
`amount:payment.amount, credit_card:`

File: service/PayTrace/PayTracePaymentServices.xml [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+                <entity-find-one entity-name="PayTrace.PaymentGatewayPayTrace" value-field="pgpt"/> 
+                <if condition="pgpt == null"><return error="true" message="No PaymentGatewayPayTrace found with ID ${paymentGatewayConfigId}"/></if>
+
+                <service-call name="PayTrace.PayTracePaymentServices.payTrace#login" in-map="context" out-map="token"/>
+
+                <script><![CDATA[
+                    org.moqui.util.RestClient restClient = ec.service.rest().method(org.moqui.util.RestClient.POST)
+                        .addHeader("Authorization", "bearer ${token.token.access_token}")
+                        .addHeader("Content-Type", "application/json").jsonObject(requestMap)
+                    restClient.uri().protocol("https").host(pgpt.transactionUrl).port(443).path(pgpt.apiVersion).path("transactions").path("sale").path("keyed").build()
```
Comment (danieltaylor-nz) [inline service/PayTrace/PayTracePaymentServices.xml]: I see no capture happening here. I think this is just auth. The documentation states
"A Capture request can be submitted later to convert this authorization into a Sale, that can be settled for payment." 

I think we need a subsequent call to capture.

File: service/PayTrace/PayTracePaymentServices.xml [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+
+                <service-call name="PayTrace.PayTracePaymentServices.payTrace#login" in-map="context" out-map="token"/>
+
+                <set field="paymentOperationEnumId" value="PgoRelease"/>
+
+                <script><![CDATA[
+                    org.moqui.util.RestClient restClient = ec.service.rest().method(org.moqui.util.RestClient.POST)
+                        .addHeader("Authorization", "bearer ${token.token.access_token}")
+                        .addHeader("Content-Type", "application/json").jsonObject(requestMap)
+                    restClient.uri().protocol("https").host(pgpt.transactionUrl).port(443).path(pgpt.apiVersion).path("transactions").path("authorization").path("capture").build()
```
Comment (danieltaylor-nz) [inline service/PayTrace/PayTracePaymentServices.xml]: this is performing another "capture", which is incorrect. I think the correct endpoint is "void"
https://developers.paytrace.com/support/home#14000044724

File: service/PayTrace/PayTracePaymentServices.xml [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+                CAPTURE        PmntDelivered
+                FORCESALE      PmntDelivered
+                SETTLED        PmntConfirmed 
+                PENDING        PmntDelivered
+                DECLINED       PmntDeclined
+                -->
+                
+                <set field="transactionStatus" from="responseMap.transaction_type"/>
+                <if condition="transactionStatus">
+                    <set field="authNetStatusMap" from="[AUTHORIZATION:'PmntAuthorized',
```
Comment (danieltaylor-nz) [inline service/PayTrace/PayTracePaymentServices.xml]: Try avoid the use of "authNet" in these services, it looks like copy paste.

File: service/PayTrace/PayTracePaymentServices.xml [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+                            paymentId:payment.paymentId, paymentMethodId:payment.paymentMethodId, referenceNum:referenceNum,
+                            amount:payment.amount, amountUomId:payment.amountUomId, transactionDate:ec.user.nowTimestamp,
+                            responseCode:responseCode, reasonCode:reasonCode, reasonMessage:reasonMessage,
+                            resultError:(responseCode ? 'N' : 'Y'), resultDeclined:(statusCode == 'N' ? 'Y' : 'N'),
+                            transactionStatus:transactionStatus]"/>
+                </if>
+            </actions>      
+        </service>
+        <service verb="get" noun="PaymentGatewayInfo">
+        <in-parameters><parameter name="paymentId" required="true"/></in-parameters>
```
Comment (danieltaylor-nz) [inline service/PayTrace/PayTracePaymentServices.xml]: I think some indentation is off here.

File: service/PayTrace/PayTracePaymentServices.xml [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+                    x_card_code:creditCard.cardSecurityCode, x_card_type:cardTypeCode,
+                    x_first_name:paymentMethod.firstNameOnAccount, x_last_name:paymentMethod.lastNameOnAccount,
+                    x_company:paymentMethod.companyNameOnAccount,
+                    x_address:postalAddress.address1, x_city:postalAddress.city, x_state:postalAddress.stateProvinceGeoId,
+                    x_zip:postalAddress.postalCode, x_country:postalAddress.countryGeoId, x_phone:phone,
+                    x_email:emailContactMech?.infoString, x_cust_id:payment.fromPartyId, x_customer_ip:visit?.clientIpAddress,
+                    x_invoice_num:payment.orderId]"/>
+            <!-- TODO, future: maybe send x_description, x_po_num -->
+        </actions>
+    </service>
```
Comment (danieltaylor-nz) [inline service/PayTrace/PayTracePaymentServices.xml]: I dont know what this service does, but it looks like A.net

File: service/PayTrace/PayTracePaymentServices.xml [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+                    <service-call name="AuthorizeDotNet.CimServices.store#CustomerPaymentMethod"
+                        in-map="[paymentMethodId:paymentMethod.paymentMethodId, paymentId:paymentId]"/>
+                    <set field="paymentMethod" from="payment.'mantle.account.method.PaymentMethod'"/>
+                </if>
+                <!-- get party after CIM check so that gatewayCimId will be populated -->
+                <set field="party" from="paymentMethod.'Owner#mantle.party.Party'"/>
+                <if condition="!party.gatewayCimId"><return message="Party ${party.partyId} has no gateway CIM ID, not processing payment"/></if>
+                <if condition="!paymentMethod.gatewayCimId"><return message="Payment Method ${paymentMethod.paymentMethodId} has no gateway CIM ID, not processing payment"/></if>
+
+                <entity-find-one entity-name="AuthorizeDotNet.PaymentGatewayAuthorizeNet" value-field="pgan"/>
```
Comment (danieltaylor-nz) [inline service/PayTrace/PayTracePaymentServices.xml]: Same as above.

**Review by danieltaylor-nz — CHANGES_REQUESTED** [review]
I think we have a lot of things to fix here.

File: service/PayTrace/PayTracePaymentServices.xml (hunk 4 above)
Comment (xlreconlx) [reply service/PayTrace/PayTracePaymentServices.xml]: Hello Daniel, this service is called "authorizeAndCapture" and it helps us to make the authorization and the capture in a single request, according to the api of Paytrace https://developers.paytrace.com/support/home#14000040854

**Comment by jonesde** [comment]
Looks good overall, merging. I do plan to rename a couple things to clarify the settings and services for storing customer data are specific to the data vault functionality and not general settings and services.

**Pull request merged**